nifi-extension-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/record/converter/RecordConverterWrapper.java

```diff
@@ -0,0 +1,78 @@
+import org.apache.nifi.util.Tuple;
+import software.amazon.kinesis.retrieval.KinesisClientRecord;
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+public class RecordConverterWrapper implements RecordConverter {
+
+    private static final String VALUE = "value";
+    private static final String METADATA = "metadata";
+
+    private static final String STREAM = "kinesis.stream";
+    private static final String SHARD_ID = "aws.kinesis.shard.id";
+    private static final String SEQUENCE_NUMBER = "aws.kinesis.sequence.number";
+    private static final String PARTITION_KEY = "aws.kinesis.partition.key";
+    private static final String APPROX_ARRIVAL_TIMESTAMP = "aws.kinesis.approximate.arrival.timestamp";
+
+    private static final RecordField FIELD_STREAM = new RecordField(STREAM, RecordFieldType.STRING.getDataType());
+    private static final RecordField FIELD_SHARD_ID = new RecordField(SHARD_ID, RecordFieldType.STRING.getDataType());
+    private static final RecordField FIELD_SEQUENCE_NUMBER = new RecordField(SEQUENCE_NUMBER, RecordFieldType.STRING.getDataType());
+    private static final RecordField FIELD_PARTITION_KEY = new RecordField(PARTITION_KEY, RecordFieldType.STRING.getDataType());
+    private static final RecordField FIELD_APPROX_ARRIVAL_TIMESTAMP = new RecordField(APPROX_ARRIVAL_TIMESTAMP, RecordFieldType.TIMESTAMP.getDataType());
+    private static final RecordSchema SCHEMA_METADATA = new SimpleRecordSchema(Arrays.asList(
+            FIELD_STREAM, FIELD_SHARD_ID, FIELD_SEQUENCE_NUMBER, FIELD_PARTITION_KEY, FIELD_APPROX_ARRIVAL_TIMESTAMP));
+
+    public static final RecordField FIELD_METADATA = new RecordField(METADATA, RecordFieldType.RECORD.getRecordDataType(SCHEMA_METADATA));
+
+
+    @Override
+    public org.apache.nifi.serialization.record.Record convert(
+            final org.apache.nifi.serialization.record.Record record,
+            final KinesisClientRecord kinesisRecord, final String streamName, final String shardId) {
+        final Tuple<String, Object> metadata = toWrapperRecordMetadata(kinesisRecord, streamName, shardId);
+        return new MapRecord(convertToWriteSchema(record.getSchema()), Map.of(metadata.getKey(), metadata.getValue(), VALUE, record));
+    }
+
+    private Tuple<String, Object> toWrapperRecordMetadata(final KinesisClientRecord consumerRecord, final String streamName, final String shardId) {
+        final Map<String, Object> metadata = new HashMap<>();
+        metadata.put(STREAM, streamName);
+        metadata.put(SHARD_ID, shardId);
+        metadata.put(SEQUENCE_NUMBER, consumerRecord.sequenceNumber());
+        metadata.put(PARTITION_KEY, consumerRecord.partitionKey());
+        final Instant approxArrivalTimestamp = consumerRecord.approximateArrivalTimestamp();
+        metadata.put(APPROX_ARRIVAL_TIMESTAMP, approxArrivalTimestamp != null ? approxArrivalTimestamp.toEpochMilli() : null);
```

Comment:
Recommend reversing the ternary conditional as follows:
```suggestion
        metadata.put(APPROX_ARRIVAL_TIMESTAMP, approxArrivalTimestamp == null ? null : approxArrivalTimestamp.toEpochMilli());
```

Reply:
NP, is there such convention explained? Always thought that the null case is less interesting and therefore it can come last

Reply:
I don't think it is written anywhere in particular, but the general approach is to evaluate the statement "positively" (if A equals B) instead of "negatively" (if A does not equal B) to make it a bit easier to follow. It isn't a strict rule, but a useful convention in some conditionals and ternaries.